### DISCUSSION TITLE
What is `System.Private.CoreLib`?

### DISCUSSION THREAD

**Discussion by colejohnson66** [dotnet/runtime#92759] [General] [answered]
What is the relationship of `System.Private.CoreLib` and `mscorlib`? Is it the replacement? If so, why the name change?

**Comment by jkotas** [comment, marked as answer]
The implementation assemblies were refactored as part of the .NET Framework to .NET Core transition. For example: In .NET Framework, the implementations of cryptography related types are spread over multiple assemblies. X509Certificate lives in mscorlib.dll and X509Certificate2 lives in System.dll that does not make sense and introduces implementation difficulties. In .NET Core, the implementation of both types lives in System.Security.Cryptography.

At the same time, .NET Core wanted to allow running .NET Framework binaries using [type forwarders](https://learn.microsoft.com/en-us/dotnet/standard/assembly/type-forwarding). .NET Core includes mscorlib.dll, System.dll, System.Core.dll, etc. These compatibility shims do not contain any code. They just forward the type from where it lives in .NET Framework to where it lives in .NET Core.

It explains why we could not just use mscorlib as the core assembly name in .NET Core and needed an assembly with a different name. We choose `System.Private.CoreLib` for the implementation assembly and `System.Runtime` for the reference assembly. The different names for implementation assembly and reference assembly were not strictly required in the hindsight, it is more an artifact of how the system evolved.

**Comment by nguyenhoanghai1605** [comment]
`System.Private.CoreLib` is the **foundational core library of .NET Core and .NET 5+**.

✅ It contains the **base types and core functionality** needed by the .NET runtime — such as:
- `System.Object`, `System.String`, `System.Int32`, `System.Exception`, etc.
- Core low-level features like threading, garbage collection interfaces, and primitive operations.

📦 **Why is it called "Private"?**  
In .NET Core, the design was changed to isolate the internal runtime-specific implementation into `System.Private.CoreLib`, rather than exposing everything publicly like in `.NET Framework`’s `mscorlib.dll`. This separation improves modularity and versioning.

💡 **How it's used:**
Even though your code might reference `System.Runtime`, `System.Collections`, etc., many of those are **just facades or forwarding assemblies**, while the actual implementation lives in `System.Private.CoreLib`.

🛠 **Built-in and loaded by the runtime**  
You don’t need to manually reference it — the runtime automatically loads it early in the app lifecycle.

📌 **Fun fact:**  
In .NET Framework, the equivalent of `System.Private.CoreLib` is `mscorlib.dll`.

---

🔍 Example:
```csharp
Console.WriteLine(typeof(object).Assembly.FullName);
// Output: System.Private.CoreLib, Version=...

**Reply by elgonzo** [reply]
Oh god, AI slop is flooding everything...
